db/db_compaction_test.cc

```diff
@@ -6464,82 +6458,366 @@ TEST_P(DBCompactionTestWithParam,
+    SetupOptions(CompactionStyle::kCompactionStyleFIFO,
+                 compaction_path_to_test);
+    DestroyAndReopen(options_);
+
+    // Create an existing SST file s0 of key range [key1,key4] and seqno range
+    // [1,2]
+    ASSERT_OK(Put("key1", "seq1"));
+    ASSERT_OK(Put("key4", "seq2"));
+    ASSERT_OK(Flush());
+    ASSERT_EQ(1, NumTableFilesAtLevel(0));
+
+    // Accumulate entries in a memtable m1 of key range [key1,key2] and seqno
+    // range [3,4] Noted that it contains a overlaped key with s0
+    ASSERT_OK(Put("key1", "seq3"));  // overlapped key
+    ASSERT_OK(Put("key2", "seq4"));
+
+    // Stop background compaction job to obtain accurate
+    // `NumTableFilesAtLevel(0)` after file ingestion and reaching
+    // `options_.level0_file_num_compaction_trigger == 3`
+    test::SleepingBackgroundTask sleeping_tasks;
+    env_->SetBackgroundThreads(1, Env::LOW);
+    env_->Schedule(&test::SleepingBackgroundTask::DoSleepTask, &sleeping_tasks,
+                   Env::Priority::LOW);
+    sleeping_tasks.WaitUntilSleeping();
+
+    // Ingested two SST files, s1 of key range [key5,key5] and seqno range [5,5]
+    // and s2 of key range [key6,key6] and seqno range [6,6]
+    IngestOneKeyValue(dbfull(), "key5", "seq5", options_);
+    IngestOneKeyValue(dbfull(), "key6", "seq6", options_);
+    // Up to now, L0 contains s0, s1, s2
+    ASSERT_EQ(3, NumTableFilesAtLevel(0));
+
+    SetupSyncPoints(compaction_path_to_test);
+    // Resume background compaction job so that Intra level0 compaction can be
+    // triggered
+    sleeping_tasks.WakeUp();
+    sleeping_tasks.WaitUntilDone();
+
+    ASSERT_OK(dbfull()->TEST_WaitForCompact());
+    ASSERT_TRUE(SyncPointsCalled())
+        << "failed for compaction path to test: " << compaction_path_to_test;
+    DisableSyncPoints();
+
+    // To verify compaction of s0, s1 and s2 (leading to new SST s4) didn't
+    // happen.
+    //
+    // Otherwise, after m1 flushes in the next step and becomes s3,
+    // we will have s3 of seqnos range [3, 4], s4 of seqnos range [1, 6]. This
+    // is a corruption because s3 is older than s4 based on largest seqno while
+    // s2 contains a newer value of "key1" than the value of "key1" contained in
+    // s4. In this case, `Get("key1")` will return an older version of the value
+    // ("seq1").
+    EXPECT_OK(Flush()) << "failed for compaction path to test: "
+                       << compaction_path_to_test;
+    EXPECT_EQ(Get("key1"), "seq3")
+        << "failed for compaction path to test: " << compaction_path_to_test;
+  }
+
+  Destroy(options_);
+}
+
+class DBCompactionTestL0FilesReorderingCorruptionWithParam
+    : public DBCompactionTestL0FilesReorderingCorruption,
+      public testing::WithParamInterface<CompactionStyle> {
+ public:
+  DBCompactionTestL0FilesReorderingCorruptionWithParam()
+      : DBCompactionTestL0FilesReorderingCorruption() {}
+};
+
+// TODO: add `CompactionStyle::kCompactionStyleLevel` to
```

Comment:
I haven't investigated but when num_level  = 1 with level compaction, there will still be a L1 in vstorage and ingested file will be placed in L1, creating difficulties to force compaction to L0. So I skipped adding UT for kCompactionStyleLevel. However let me know if you still see value in investigating this and figure this out!

Reply:
I think we sanitize num_levels to be at least 3 or something in level compaction.

Reply:
Oh okay thanks! Then CompactRangeOptions::change_level=true and CompactRangeOptions::target_level=0 might help with the unit test design. Will try